dmic: add the correct gain values to the dmic registers

fkwasowi opened the pull request:
Zephyr increments the gain until it reaches the maximum value and then sets the registers to zero which is incorrect. The values set in the DMIC config should be restored.

dabekjakub reviewed (commented):
What about gain bypass ? Comment lines above states that dmic gain value 0 has a purpose. This comment or the behaviour should be changed. If comment is deprecated it should be changed. Also i see no correction for dmic->gain if the gval = 0. Should this value be set to max as in the if above this change ?

fkwasowi commented:
SOF CI TEST

fkwasowi commented:
> What about gain bypass ? Comment lines above states that dmic gain value 0 has a purpose. This comment or the behaviour should be changed. If comment is deprecated it should be changed. Also i see no correction for dmic->gain if the gval = 0. Should this value be set to max as in the if above this change ?

Setting gain to zero is incorrect behaviour. Instead, we should restore the values we sent in the Copier DMIC config. If zero is the expected value (sent in the config) this will also be taken into account.

fkwasowi committed:
dmic: add the correct gain values to the dmic registers

Zephyr increments the gain until it reaches the maximum value and
then sets the registers to zero which is incorrect.
The values set in the DMIC config should be restored.

Signed-off-by: Fabiola Kwasowiec <fabiola.kwasowiec@intel.com>

singalsu commented:
Yes, the change is correct, good catch @fkwasowi ! Zero is a shortcut to gain bypass when gain is exactly 1 (that is not a possible Q1.19 value).

The gain ramp in capture beginning should be also fixed to stop at the value in the blob. I wonder if the blob had e.g. -6 dB as gain, the ramp would go up to 0 dB and then jump to -6 dB. Can you please check that as well.

fkwasowi commented:
> Yes, the change is correct, good catch @fkwasowi ! Zero is a shortcut to gain bypass when gain is exactly 1 (that is not a possible Q1.19 value).
> 
> The gain ramp in capture beginning should be also fixed to stop at the value in the blob. I wonder if the blob had e.g. -6 dB as gain, the ramp would go up to 0 dB and then jump to -6 dB. Can you please check that as well.

Tests take this case into account - for example the TestDmicFeatures.test_107_03_ace_generic_dmic_gain_left_right[L-6dB] where we get results consistent with the expected.

singalsu commented:
> > Yes, the change is correct, good catch @fkwasowi ! Zero is a shortcut to gain bypass when gain is exactly 1 (that is not a possible Q1.19 value).
> > The gain ramp in capture beginning should be also fixed to stop at the value in the blob. I wonder if the blob had e.g. -6 dB as gain, the ramp would go up to 0 dB and then jump to -6 dB. Can you please check that as well.
> 
> Tests take this case into account - for example the TestDmicFeatures.test_107_03_ace_generic_dmic_gain_left_right[L-6dB] where we get results consistent with the expected.

Thanks, all should be good then!
